File: to_Danya/ru-olymp-regional-2021-archive/day2/8/solutions/a-plus-b_mp_group2.cpp
#include <bits/stdc++.h>
#define sz(x) ((int)(x).size())

const int MOD = 1000000007;

using namespace std;

int main() {
  string a, b, c;
  cin >> a >> b >> c;

  int n = sz(a);
  vector<int> createCarry(n), needCarry(n);
  for (int i = 0; i < n; ++i) {
    int x = a[i] - '0';
    int y = b[i] - '0';
    int z = c[i] - '0';
    needCarry[i] = (x + y) % 10 != z;
    createCarry[i] = x + y + needCarry[i] >= 10;
    if (x + y + needCarry[i] - createCarry[i] * 10 != z) {
      cout << "0\n";
      return 0;
    }
  }

  vector<vector<int>> d(n, vector<int>(1 << n, 0));

  for (int mask = 1; mask < 1 << n; ++mask)
    for (int v = 0; v < n; ++v)
      if (mask & (1 << v)) {
        int pmask = mask ^ (1 << v);
        int &cd = d[v][mask];
        if (pmask == 0)
          cd = a[v] != '0' && b[v] != '0' && c[v] != '0' && !createCarry[v];
        else
          for (int pv = 0; pv < n; ++pv)
            if (pmask & (1 << pv))
              if (createCarry[v] == needCarry[pv])
                (cd += d[pv][pmask]) %= MOD;
      }

  int ans = 0;
  for (int i = 0; i < n; ++i)
    if (!needCarry[i]) 
      (ans += d[i][(1 << n) - 1]) %= MOD;

  cout << ans << "\n";
}

File: to_Danya/ru-olymp-regional-2021-archive/day2/8/solutions/a-plus-b_mp_group2.java
import java.util.Scanner;

public class APlusB2 {
  private static final int MOD = 1000000007;

  public int solve(String a, String b, String c) {
    int n = a.length();
    boolean[] createCarry = new boolean[n];
    boolean[] needCarry = new boolean[n];
    for (int i = 0; i < n; ++i) {
      int x = a.charAt(i) - '0';
      int y = b.charAt(i) - '0';
      int z = c.charAt(i) - '0';
      needCarry[i] = (x + y) % 10 != z;
      createCarry[i] = x + y + (needCarry[i] ? 1 : 0) >= 10;
      if (x + y + (needCarry[i] ? 1 : 0) - (createCarry[i] ? 10 : 0) != z) {
        return 0;
      }
    }

    int[][] d = new int[n][1 << n];

    for (int mask = 1; mask < 1 << n; ++mask)
      for (int v = 0; v < n; ++v)
        if ((mask & (1 << v)) != 0) {
          int pmask = mask ^ (1 << v);
          if (pmask == 0)
            d[v][mask] = a.charAt(v) != '0' && b.charAt(v) != '0' && c.charAt(v) != '0' && !createCarry[v] ? 1 : 0;
          else
            for (int pv = 0; pv < n; ++pv)
              if ((pmask & (1 << pv)) != 0)
                if (createCarry[v] == needCarry[pv])
                  d[v][mask] = (d[v][mask] + d[pv][pmask]) % MOD;
        }

    int ans = 0;
    for (int i = 0; i < n; ++i)
      if (!needCarry[i])
        ans = (ans + d[i][(1 << n) - 1]) % MOD;
    return ans;
  }

  public static void main(String[] arg) {
    Scanner scanner = new Scanner(System.in);
    String a = scanner.next();
    String b = scanner.next();
    String c = scanner.next();
    System.out.println(new APlusB2().solve(a, b, c));
  }
}
